### ISSUE TITLE
renderD128 is not accessible in the container

### ISSUE THREAD

**Issue by GoingTime** [jellyfin/jellyfin#9229]
**FFmpeg Logs**
```shell
 ffmpeg version 5.1.2-Jellyfin Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 10 (Debian 10.2.1-6)
  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-libs=-lfftw3f --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-shared --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto --enable-gpl --enable-version3 --enable-static --enable-gmp --enable-gnutls --enable-chromaprint --enable-libdrm --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libdav1d --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=amd64 --enable-libsvtav1 --enable-libshaderc --enable-libplacebo --enable-vulkan --enable-opencl --enable-vaapi --enable-amf --enable-libmfx --enable-ffnvcodec --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvdec --enable-nvenc
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
[AVHWDeviceContext @ 0x55abe4731e80] No VA display found for any default device.
Device creation failed: -22.
Failed to set value 'vaapi=va:,driver=iHD,kernel_driver=i915' for option 'init_hw_device': Invalid argument
Error parsing global options: Invalid argument
```

**Server System:**
 - OS: [TrueNAS-SCALE-22.12.0 Base Debian 11]
 - Jellyfin Version: [10.8.9]
 - Hardware settings & device: [Intel i5 10500]
 - Reverse proxy: [nginx]
 
**Client System:**
 - OS: [Windows]
 - Client: [Web/Browser]
 - Browser: [Microsoft Edge]
 - Client and Browser Version: [108.0.1462.54 64bit]]

Jellyfin uses Docker for deployment, does not enable additional CAP permissions, runAsUser uses ordinary users, and adds video and render additional groups.

```shell
#### in the container ####
$ whoami
whoami: cannot find name for user ID 1234
$ cat /dev/dri/renderD128
cat: /dev/dri/renderD128: Permission denied
$ groups
groups: cannot find name for group ID 1234
1234
$ ls -la /dev/dri
total 0
drwxr-xr-x 2 root root        80 Feb  1 16:15 .
drwxr-xr-x 6 root root       380 Feb  1 16:15 ..
crw-rw---- 1 root video 226,   0 Feb  1 16:15 card0
crw-rw---- 1 root   107 226, 128 Feb  1 16:15 renderD128
```

```shell
#### in the host ####
host% whoami
test
host% cat /dev/dri/renderD128
^C # I can access card0 & renderD128
host% groups
myapp video render builtin_users apps
host% ls -la /dev/dri
total 0
drwxr-xr-x  3 root root        100 Feb  1 22:56 .
drwxr-xr-x 21 root root       3960 Feb  1 22:57 ..
drwxr-xr-x  2 root root         80 Feb  1 22:56 by-path
crw-rw----  1 root video  226,   0 Feb  1 22:56 card0
crw-rw----  1 root render 226, 128 Feb  1 22:56 renderD128
```

In short, with the same user, I can access renderD128 normally in the host machine, but not in the container, unless I use the root user. Even if I enable privileged mode, do device mapping, or add CAP_SYS_RAWIO permission, I cannot access renderD128 in the container.

**Comment by cvium** [comment]
Are you sure this is a Jellyfin issue? We do not support truenas scale.

**Comment by poulpor** [comment]
Not sure it's a jellyfin problem, it's dirty, but did you try to change permissions of  card0 and renderD128 ? Something like:
chmod 666 /dev/dri/card0
chmod 666 /dev/dri/renderD128

**Comment by poulpor** [comment]
Just create a startup script.

```
#!/bin/bash
#NAS Startup Script
#Access for HW Decoder Jelllyfin
chmod 666 /dev/dri/card0
chmod 666 /dev/dri/renderD128
exit
```

**Comment by poulpor** [comment]
I don't use unraid but there seems to be a "User scripts plugin", then schedule it at "array starting"..?

**Comment by chteow** [comment]
you access docker with user mode. 
Try root to see
add /bin/bash at the end to access root

**Comment by mastercko** [comment]
I had a similar problem show up and mess up my Jellyfin install (I have it in a docker on a spare Ubuntu box in my network closet). After several days of searching, this and similar threads finally pointed me to the answer, which is that the user within the Jellyfin container does not have the proper group membership to access the devices, even though it does on the host.

Because giving all users rw permissions (chmod 666) to those drivers is not good security practice (same with running your container as root), and also it unsets on reboot, the clean solution was simply adding the `video (44)` and `render (110)` groups to my docker compose definition for Jellyfin. 

I think technically, you only need to add the `render` group to get access to the the `renderD128` device, but just to be safe, I added both groups since I have given the container access to both the `renderD128` and `card0` drivers. Whether you use a standard docker run command or a docker compose definition, if you just do a `group_add` with those two groups, the devices should be accessible and your hardware rendering should be good to go.

In case it helps anyone, here is my (sanitized) compose yaml:
```
---
version: "3"
services:
  jellyfin:
    image: jellyfin/jellyfin:latest
    container_name: jellyfin
    user: 1000:1000                             # main user, group
    group_add:
      - 44                                      # add video group
      - 110                                     # add render group
    volumes:
      - jellyfin-config:/config
      - jellyfin-cache:/cache
      - /mnt/media_share:/media                 # FYI: mounted via NAS cifs share
      - /etc/passwd:/etc/passwd:ro              # import user names from host
      - /etc/group:/etc/group:ro                # import group names from host
    network_mode: host
    restart: unless-stopped
    privileged: true
    devices:                                    # import drivers for HW accel
      - /dev/dri/card0:/dev/dri/card0
      - /dev/dri/renderD128:/dev/dri/renderD128
volumes:
  jellyfin-config:
    external: true
  jellyfin-cache:
    external: true
```

**Comment by glassman81** [comment]
>     devices:                                    # import drivers for HW accel
>       - /dev/dri/card0:/dev/dri/card0
>       - /dev/dri/renderD128:/dev/dri/renderD128
> ```

Wouldn't '- /dev/dri:/dev/dri' be good enough as a single line to add both?? Why do they have to be added separately?

**Comment by mastercko** [comment]
Oh, yeah, that's a good point, I suppose that would work ok. To be honest, I just didn't think about it because I was adding them one at a time to test.

**Comment by DocMAX** [comment]
```
root@042a7b27a7b8:/# ls -la /dev/dri/
total 0
drwxrwxrwx 4 root root       120 Feb 10 23:59 .
drwxr-xr-x 6 root root       360 Feb 11 01:11 ..
drwxr-xr-x 2 root root        80 Feb  9 12:08 by-path
drwxr-xr-x 2 root root        40 Feb 10 23:59 card0
crwxrwxrwx 1 root video 226,   1 Feb  9 12:08 card1
crwxrwxrwx 1 root   106 226, 128 Feb  9 12:08 renderD128
root@042a7b27a7b8:/# cat /dev/dri/renderD128
cat: /dev/dri/renderD128: Operation not permitted
```
Same problem here. And chmod 666 + restart also doesn't work. but i have other containers on the same machine where access works.
